File: 4D_Recon_Launcher_for_Real_Data/XT_FBP_Reconstruction.py
import numpy as np
from XT_IOMisc import create_folder, convert_um2HU, convert_HU2uint8, write_array2tif
from skimage.transform import iradon
from math import floor

def gen_centered_projections (projections, rotation_center):
	centered_proj_sz = 2*rotation_center+1
	centered_projections = np.zeros((projections.shape[0], centered_proj_sz, projections.shape[2]), dtype = np.float64, order = 'C') 
	centered_projections[:, 0:min(centered_proj_sz,projections.shape[1]), :] = projections[:, 0:min(centered_proj_sz,projections.shape[1]), :]
	return centered_projections

def do_FBP_reconstruction (proj, recon, files):

########################### CHANGE RESULT FOLDER BELOW #########################
	result_folder = files['Result_Folder']
################################################################################

	for k in range(len(recon['Rtime_num'])):		
		path2launch = files['Launch_Folder'] + 'run_' + 'sigs_' + str(recon['sigma_s'][k]) + '_sigt_' + str(recon['sigma_t'][k]) + '_r_' + str(recon['r'][k]) + '_K_' + str(proj['K']) + '_N_theta_' + str(proj['N_theta'])  + '_N_p_' + str(proj['recon_N_p']) + recon['msg_string'] + '/'
		projections = np.zeros((proj['recon_N_p'], proj['recon_N_r'], proj['recon_N_t']), dtype = np.float64, order = 'C')
		ratio = proj['recon_N_t']/recon['node_num']
		for rank in range(recon['node_num']):
			projections[:,:,rank*ratio:(rank+1)*ratio] = np.fromfile(path2launch + 'projection_n' + str(rank) + '.bin', dtype=np.float64, count=-1).reshape((proj['recon_N_p'],proj['recon_N_r'],ratio),order='C') 	
		centered_projections = gen_centered_projections(projections,floor(proj['rotation_center_r']))
		path2results = result_folder + 'FBP_' + 'r_' + str(recon['r'][k]) + '_K_' + str(proj['K']) + '_N_theta_' + str(proj['N_theta']) + '_N_p_' + str(proj['recon_N_p']) + recon['msg_string'] + '/'
		create_folder(path2results)
		
		for i in range(int(recon['Rtime_num'][k])):
			time_start = recon['Rtime0'] + i*recon['Rtime_delta'][k]
			time_stop = recon['Rtime0'] + (i + 1)*recon['Rtime_delta'][k]
			indices = np.bitwise_and(proj['times'] >= time_start, proj['times'] < time_stop)
			for k in range(recon['node_num']):
				ratio = proj['recon_N_t']/recon['node_num']
				idx = 0
				Obj2Bin = np.zeros((ratio, proj['recon_N_r'], proj['recon_N_r']), dtype=np.float64, order='C')
				for j in range(k*ratio, (k+1)*ratio):
					Object = iradon (np.transpose(centered_projections[indices,:, j]), 90-np.mod(proj['angles'][indices],360), output_size=proj['recon_N_r'], filter="hamming", interpolation='linear')/(proj['length_r']/proj['recon_N_r'])
					#Object = iradon (np.transpose(centered_projections[indices,:]), np.mod(proj['angles'][indices],180), output_size=centered_projections.shape[1], filter="hamming", interpolation='linear')/(proj['length_r']/proj['N_r'])
					Object = np.transpose(Object)
					Obj2Bin[idx,:,:] = Object
					idx = idx + 1
					Object = convert_um2HU(Object)
					Object_uint8 = convert_HU2uint8(Object, recon['maxHU'], recon['minHU'])
					write_array2tif(path2launch + 'object_time' + str(i) + '_z_' + str(j) + '.tif', Object_uint8)
				fid = open(path2launch + 'object_n' + str(k) + '_time_' + str(i)  + '.bin','wb')
				Obj2Bin.tofile(fid)
				fid.close()		

File: 4D_Recon_Launcher_for_Real_Data/XT_FBP_Reconstruction.m
clear; close all;

K = 8;
r = 8;
N_theta = 256;
N_p = 256*4;
N_r = 256;
N_t = 4;
rotation_center = 131;
node_num = 1;
length_r = 0.65*1024;
path2result = '../../Recon_Runs/Att_Recon_Sim_256/XT_Result_Repository/FBP_r_8_K_8_N_theta_256_N_p_1024/';
path2launch = '../../Recon_Runs/Att_Recon_Sim_256/XT_run/run_sigs_300000.0_sigt_3000.0_r_8_K_8_N_theta_256_N_p_1024_sig2_10_th0.1/';
L = N_theta/K;
[views, buf1, buf2] = gen_interlaced_views_0_to_inf (L, K, N_p);

Rtime_num = r*N_p/N_theta;
Rtime_delta = N_p/Rtime_num;

ratio = N_t/node_num;
centered_proj_sz = 2*rotation_center + 1;
x_idx = 1:min(N_r,centered_proj_sz);
y_idx = (1:N_r) + floor((centered_proj_sz-N_r)/2);
for rank = 1:node_num
    fid = fopen([path2launch,'projection_n', num2str(rank-1), '.bin'], 'r');
    proj_temp = fread(fid, N_p*N_r*ratio,'double');
    fclose(fid);
    projections = permute(reshape(proj_temp,[ratio,N_r,N_p]),[2,3,1]); 
    centered_projections = zeros(centered_proj_sz, N_p, ratio); 
    centered_projections(1:min(centered_proj_sz,N_r), :, :) = projections(1:min(centered_proj_sz,N_r), :, :);

    recon = zeros(N_r, N_r, ratio); 
    for i = 1:Rtime_num
        for j = 1:ratio
            temp = iradon(centered_projections(:,(i-1)*N_theta/r+1:i*N_theta/r,j), views((i-1)*N_theta/r+1:i*N_theta/r), 'cubic', 'Hamming', 1, centered_proj_sz);
            recon(:,1:min(N_r,centered_proj_sz),j) = N_r/length_r*temp(y_idx,1:min(N_r,centered_proj_sz));
        end
        recon(recon < 0) = 0;
        fid = fopen([path2launch, 'object_n', num2str(rank-1), '_time_', num2str(i-1), '.bin'], 'wb');
        fwrite(fid, permute(recon,[2,1,3]), 'double');
        fclose(fid);    
    end
end
